File: python/src/azoth/process/reference/heat_exchanger.py
"""``process.heat_exchanger`` - the exchanger's kernel.

Spec: ``specs/models/process/heat_exchanger.toml``

The Python twin of ``crates/azoth-process/src/models/heat_exchanger.rs`` over
``crates/azoth-process/src/kernels/heat_exchanger.rs``, written to mirror them.

# The rating, not a duty

``HeatExchanger.run``'s default branch specifies ``UAvalue`` and runs an effectiveness-NTU
rating: each side's capacity is estimated by flashing it at the **other side's inlet
temperature** and dividing the total enthalpy change by that temperature change, then
``NTU = UA / C_min`` and ``calcThermalEffectivenes(NTU, C_min / C_max)`` scales that swing.
``duty`` is an *output* of this class - ``energyInput`` appears nowhere in it - so a kernel
that moved a caller's duty was a different unit operation under the same name.

``C`` is a difference of *totals* over a difference of temperatures, which is W/K; that is
what makes the NTU dimensionless. The absolute value matters: the hot side's seeded swing
is negative because it cools, and taking it at face value would make ``C_min`` and the NTU
negative, which is an exponential in the wrong direction.

At an effectiveness of one ``run`` leaves the capacity-limiting side at its seeded state -
"this side at the other's inlet temperature". That is the same *enthalpy* the arithmetic
gives, so the branch is an optimisation rather than a second answer, and this reaches it
through the flash either way.

# The other mode pins an outlet

``hot_outlet_temperature`` or ``cold_outlet_temperature`` pins that side at its inlet
pressure and energy-balances the other against it. **NeqSim's own ``runSpecifiedStream``
does not reach the state it names**: it clones the stream's already-flashed fluid, sets the
temperature and flashes again, and the enthalpy it lands on is one NeqSim's own ``PHflash``
places at another temperature. This reaches the pinned temperature directly, which is why
the case is pinned to the capture's ``reference_pin_*`` rows rather than to its
``out_temperature_*`` ones.
"""

from __future__ import annotations

import math

from azoth.core.errors import InvalidInputError
from azoth.core.range import apply_checks, checks_for
from azoth.core.result import HeatExchangerResult
from azoth.core.units import Q, from_si, input_to_si
from azoth.core.warnings import Warning
from azoth.eos import components as _components
from azoth.eos.reference.ph_flash import enthalpy_at
from azoth.eos.reference.ph_flash import ph_flash as ph_flash_solve

#: The arrangements ``calcThermalEffectivenes`` names, in azoth's own spelling.
FLOW_ARRANGEMENTS = ("counterflow", "parallelflow", "shell_and_tube")


def heat_exchanger(
    hot_components: list[str],
    hot_in_n: Q,
    hot_in_z: list[float],
    hot_in_p: Q,
    hot_in_t: Q,
    cold_components: list[str],
    cold_in_n: Q,
    cold_in_z: list[float],
    cold_in_p: Q,
    cold_in_t: Q,
    ua: Q | None = None,
    flow_arrangement: str = "counterflow",
    hot_outlet_temperature: Q | None = None,
    cold_outlet_temperature: Q | None = None,
) -> HeatExchangerResult:
    """Exchange heat between two streams.

    Args:
        hot_components: the hot side's substances, by name.
        hot_in_n: the hot inlet's molar flow.
        hot_in_z: the hot inlet's composition.
        hot_in_p: the hot inlet's pressure.
        hot_in_t: the hot inlet's temperature.
        cold_components: the cold side's substances, by name.
        cold_in_n: the cold inlet's molar flow.
        cold_in_z: the cold inlet's composition.
        cold_in_p: the cold inlet's pressure.
        cold_in_t: the cold inlet's temperature.
        ua: the overall conductance, for the rating. Unread when an outlet is pinned.
        flow_arrangement: ``counterflow``, ``parallelflow`` or ``shell_and_tube``.
        hot_outlet_temperature: pin the hot outlet at this temperature.
        cold_outlet_temperature: pin the cold outlet at this temperature.

    Returns:
        Both outlets' records.

    Raises:
        InvalidInputError: for the mode and arrangement refusals the spec lists.

    Example:
        >>> import azoth
        >>> q = azoth.ureg.Quantity
        >>> r = heat_exchanger(
        ...     ["methane", "n-butane"],
        ...     q(1.0, "mol/s"),
        ...     [0.8, 0.2],
        ...     q(20.0, "bar"),
        ...     q(400.0, "K"),
        ...     ["n-butane", "n-pentane"],
        ...     q(1.0, "mol/s"),
        ...     [0.5, 0.5],
        ...     q(5.0, "bar"),
        ...     q(300.0, "K"),
        ...     q(100.0, "W/K"),
        ...     "counterflow",
        ... )
        >>> round(r.hot_out_t.to("K").magnitude, 4)
        313.7571
    """
    spec = _spec()
    checks = checks_for(spec)
    warnings: list[Warning] = []

    ua_si = None if ua is None else input_to_si(spec, "ua", ua)
    hot_t_si = input_to_si(spec, "hot_in_t", hot_in_t)
    cold_t_si = input_to_si(spec, "cold_in_t", cold_in_t)
    # Every declared bound's quantity is fed here: a closure that supplies some and not
    # the rest emits a `RANGE_CHECK_SKIPPED` warning for a check that could have run.
    apply_checks(
        checks.on_input,
        {"ua": ua_si, "hot_in_t": hot_t_si, "cold_in_t": cold_t_si}.get,
        warnings,
    )

    if hot_outlet_temperature is not None and cold_outlet_temperature is not None:
        raise InvalidInputError(
            "hot_outlet_temperature",
            "an exchanger pins one outlet at most: with both stated the other side has "
            "nothing left to solve for",
        )
    pinned = hot_outlet_temperature is not None or cold_outlet_temperature is not None
    if not pinned and ua_si is None:
        raise InvalidInputError(
            "ua",
            "the rating needs an overall conductance. NeqSim's `UAvalue` defaults to "
            "500 W/K, and this refuses to guess it: a default that only changes the "
            "answer is not one a caller can see",
        )
    if flow_arrangement not in FLOW_ARRANGEMENTS:
        raise InvalidInputError(
            "flow_arrangement",
            f"{flow_arrangement} is not one of {list(FLOW_ARRANGEMENTS)}. NeqSim's `run` "
            f"falls through to the counterflow relation for an arrangement it does not "
            f"know, which would make a misspelling a plausible answer",
        )

    hot_mixture, hot_gas = _components.mixture_of(hot_components, eos="pr")
    cold_mixture, cold_gas = _components.mixture_of(cold_components, eos="pr")
    hot_n = input_to_si(spec, "hot_in_n", hot_in_n)
    cold_n = input_to_si(spec, "cold_in_n", cold_in_n)
    hot_p = input_to_si(spec, "hot_in_p", hot_in_p)
    cold_p = input_to_si(spec, "cold_in_p", cold_in_p)
    hot_h, _ = enthalpy_at(hot_mixture, hot_gas, hot_t_si, hot_p, hot_in_z)
    cold_h, _ = enthalpy_at(cold_mixture, cold_gas, cold_t_si, cold_p, cold_in_z)

    if pinned:
        hot_pinned = hot_outlet_temperature is not None
        pinned_t = input_to_si(
            spec,
            "hot_outlet_temperature" if hot_pinned else "cold_outlet_temperature",
            hot_outlet_temperature if hot_pinned else cold_outlet_temperature,
        )
        if hot_pinned:
            specified, _ = enthalpy_at(hot_mixture, hot_gas, pinned_t, hot_p, hot_in_z)
            duty = hot_n * (specified - hot_h)
            hot_out_h, cold_out_h = specified, cold_h - duty / cold_n
        else:
            specified, _ = enthalpy_at(cold_mixture, cold_gas, pinned_t, cold_p, cold_in_z)
            duty = cold_n * (specified - cold_h)
            cold_out_h, hot_out_h = specified, hot_h - duty / hot_n
    else:
        span = abs(hot_t_si - cold_t_si)
        # The same guard the Rust uses, so the two refuse the same inputs rather than
        # agreeing by accident on the cases that do not reach it.
        if span <= 2.220446049250313e-16 * max(abs(hot_t_si), abs(cold_t_si)):
            raise InvalidInputError(
                "hot_in_t",
                f"both inlets are at {hot_t_si} K, so there is no driving force for the "
                f"rating to size against",
            )
        assert ua_si is not None  # the mode check above guarantees it
        hot_seeded, _ = enthalpy_at(hot_mixture, hot_gas, cold_t_si, hot_p, hot_in_z)
        cold_seeded, _ = enthalpy_at(cold_mixture, cold_gas, hot_t_si, cold_p, cold_in_z)
        hot_swing = hot_n * (hot_seeded - hot_h)
        cold_swing = cold_n * (cold_seeded - cold_h)
        c_min = min(abs(hot_swing), abs(cold_swing)) / span
        c_max = max(abs(hot_swing), abs(cold_swing)) / span
        effectiveness = _effectiveness(ua_si / c_min, c_min / c_max, flow_arrangement)
        # The side whose seeded swing is the larger is the one `run` keeps, and the other
        # is energy-balanced against it.
        if abs(cold_swing) > abs(hot_swing):
            duty = effectiveness * hot_swing
            hot_out_h, cold_out_h = hot_h + duty / hot_n, cold_h - duty / cold_n
        else:
            duty = effectiveness * cold_swing
            cold_out_h, hot_out_h = cold_h + duty / cold_n, hot_h - duty / hot_n

    return HeatExchangerResult(
        hot_out_n=from_si(hot_n, "mol/s"),
        hot_out_z=tuple(hot_in_z),
        hot_out_p=hot_in_p,
        hot_out_t=_temperature(hot_mixture, hot_gas, hot_p, hot_out_h, hot_in_z),
        hot_out_h=from_si(hot_out_h, "J/mol"),
        cold_out_n=from_si(cold_n, "mol/s"),
        cold_out_z=tuple(cold_in_z),
        cold_out_p=cold_in_p,
        cold_out_t=_temperature(cold_mixture, cold_gas, cold_p, cold_out_h, cold_in_z),
        cold_out_h=from_si(cold_out_h, "J/mol"),
        warnings=tuple(warnings),
    )


def _temperature(
    mixture: object, ideal_gas: object, pressure: float, h: float, z: list[float]
) -> Q:
    """The temperature a side reaches at a shifted enthalpy."""
    solved = ph_flash_solve(
        mixture,  # type: ignore[arg-type]
        ideal_gas,  # type: ignore[arg-type]
        from_si(pressure, "Pa"),
        from_si(h, "J/mol"),
        z,
    )
    return solved.T


def _effectiveness(ntu: float, capacity_ratio: float, arrangement: str) -> float:
    """The effectiveness ``calcThermalEffectivenes`` carries, for the three arrangements."""
    if capacity_ratio == 0.0:
        return 1.0 - math.exp(-ntu)
    if arrangement == "parallelflow":
        return (1.0 - math.exp(-ntu * (1.0 + capacity_ratio))) / (1.0 + capacity_ratio)
    if arrangement == "counterflow" and capacity_ratio == 1.0:
        return ntu / (1.0 + ntu)
    exp = math.exp(-ntu * (1.0 - capacity_ratio))
    return (1.0 - exp) / (1.0 - capacity_ratio * exp)


def _spec() -> dict[str, object]:
    """The generated spec, imported here so the module has no import-time cycle."""
    from azoth._models_gen import model

    return model("process.heat_exchanger")

File: crates/azoth-process/src/models/heat_exchanger.rs
//! `process.heat_exchanger` - the exchanger's kernel as a registered id.
//!
//! Spec: `specs/models/process/heat_exchanger.toml`. The arithmetic is
//! [`crate::kernels::heat_exchanger`]; what is here is the boundary a case and a
//! cross-impl test address.
//!
//! **The first id whose two ports carry different fluids.** Every other unit operation
//! declares one `components`; this declares `hot_components` and `cold_components`, and
//! the boundary derives the two `Stream`s from them separately.

use azoth_core::units::{
    MolarEnergy, Power, Pressure, ThermalConductance, ThermodynamicTemperature, joules_per_mole,
};
use azoth_core::{CalcResult, Result, Warning, apply_checks};
use serde::Serialize;

use crate::executor::json::{optional_scalar, scalar, warnings as wire_warnings};
use crate::kernels::heat_exchanger as kernel;
use crate::model_gen;
use crate::stream::Stream;

/// Result of `process.heat_exchanger`.
///
/// Two single-multiplicity ports, so the record's five fields per port - the same shape
/// `process.separator` has, under this unit operation's own port names - and the duty and the
/// rating's own numbers beside them.
///
/// **Serialisable, so that a flowsheet's run publishes it.** The duty is on no outlet stream, and
/// neither are the four numbers a rating sized the exchanger by; the wire writes each through
/// [`crate::executor::json::scalar`] and writes `null` for the rating's where there was no rating
/// to size one, which is the same absence a flash's single phase is.
#[derive(Debug, Clone, PartialEq, Serialize)]
pub struct HeatExchangerResult {
    /// Hot outlet molar flow, mol/s.
    pub hot_out_n: f64,
    /// Hot outlet composition.
    pub hot_out_z: Vec<f64>,
    #[serde(serialize_with = "scalar")]
    /// Hot outlet pressure.
    pub hot_out_p: Pressure,
    #[serde(serialize_with = "scalar")]
    /// Hot outlet temperature.
    pub hot_out_t: ThermodynamicTemperature,
    #[serde(serialize_with = "scalar")]
    /// Hot outlet molar enthalpy.
    pub hot_out_h: MolarEnergy,
    /// Cold outlet molar flow, mol/s.
    pub cold_out_n: f64,
    /// Cold outlet composition.
    pub cold_out_z: Vec<f64>,
    #[serde(serialize_with = "scalar")]
    /// Cold outlet pressure.
    pub cold_out_p: Pressure,
    #[serde(serialize_with = "scalar")]
    /// Cold outlet temperature.
    pub cold_out_t: ThermodynamicTemperature,
    #[serde(serialize_with = "scalar")]
    /// Cold outlet molar enthalpy.
    pub cold_out_h: MolarEnergy,
    /// The heat the hot side released, W: positive when the hot side cools, which is what the
    /// cold side's gain balances. The one figure that means the same thing in both modes.
    #[serde(serialize_with = "scalar")]
    pub duty: Power,
    /// `UA / C_min`, dimensionless, or `null` where one outlet was pinned instead of rated.
    pub ntu: Option<f64>,
    /// The swing the relation scaled, or `null` where one outlet was pinned.
    pub effectiveness: Option<f64>,
    /// The smaller of the two estimated capacities, W/K, or `null` where one outlet was pinned.
    #[serde(serialize_with = "optional_scalar")]
    pub c_min: Option<ThermalConductance>,
    /// The larger of them, W/K, or `null` where one outlet was pinned.
    #[serde(serialize_with = "optional_scalar")]
    pub c_max: Option<ThermalConductance>,
    /// `C_min / C_max`, dimensionless, or `null` where one outlet was pinned.
    pub capacity_ratio: Option<f64>,
    /// Caveats.
    #[serde(serialize_with = "wire_warnings")]
    pub warnings: Vec<Warning>,
}

impl CalcResult for HeatExchangerResult {
    const CALC_ID: &'static str = "process.heat_exchanger";
    const FIELDS: &'static [&'static str] = &[
        "hot_out_n",
        "hot_out_z",
        "hot_out_p",
        "hot_out_t",
        "hot_out_h",
        "cold_out_n",
        "cold_out_z",
        "cold_out_p",
        "cold_out_t",
        "cold_out_h",
        "duty",
        "ntu",
        "effectiveness",
        "c_min",
        "c_max",
        "capacity_ratio",
        "warnings",
    ];

    fn warnings(&self) -> &[Warning] {
        &self.warnings
    }
}

/// Exchange heat between two streams.
///
/// # Errors
/// [`azoth_core::AzothError::InvalidInput`] for the mode and arrangement refusals
/// [`crate::kernels::heat_exchanger`] lists, and whatever the databank refuses.
#[allow(clippy::too_many_arguments)] // one parameter per declared input, and there are fourteen
pub fn heat_exchanger(
    hot_components: &[String],
    hot_in_n: f64,
    hot_in_z: &[f64],
    hot_in_p: Pressure,
    hot_in_t: ThermodynamicTemperature,
    cold_components: &[String],
    cold_in_n: f64,
    cold_in_z: &[f64],
    cold_in_p: Pressure,
    cold_in_t: ThermodynamicTemperature,
    ua: Option<ThermalConductance>,
    flow_arrangement: &str,
    hot_outlet_temperature: Option<ThermodynamicTemperature>,
    cold_outlet_temperature: Option<ThermodynamicTemperature>,
) -> Result<HeatExchangerResult> {
    let spec = &model_gen::HEAT_EXCHANGER_SPEC;
    let mut warnings = Vec::new();
    // One call over every declared input, for the reason `reactions.chemical_equilibrium`
    // gives. `ua` is optional, so an absent one is a *skipped check* rather than a pass -
    // which is the whole reason the bound is declared on it and not on the temperatures.
    apply_checks(
        spec.input_checks(),
        |quantity| match quantity {
            "ua" => ua.map(|value| value.value),
            "hot_in_t" => Some(hot_in_t.value),
            "cold_in_t" => Some(cold_in_t.value),
            _ => None,
        },
        &mut warnings,
    )?;

    let hot = Stream::from_pt(
        hot_components.to_vec(),
        hot_in_z.to_vec(),
        hot_in_n,
        hot_in_p,
        hot_in_t,
    )?;
    let cold = Stream::from_pt(
        cold_components.to_vec(),
        cold_in_z.to_vec(),
        cold_in_n,
        cold_in_p,
        cold_in_t,
    )?;
    let outcome = kernel(
        &hot,
        &cold,
        ua,
        flow_arrangement,
        hot_outlet_temperature,
        cold_outlet_temperature,
    )?;
    Ok(HeatExchangerResult::of(&outcome, warnings))
}

impl HeatExchangerResult {
    /// The result of one kernel call, as a flowsheet's dispatcher reads it.
    ///
    /// **The warnings are the caller's, because the two callers have different ones.** A case runs
    /// the spec's input checks through [`apply_checks`]; a flowsheet's equivalent is the
    /// *checker's*, which reports through the envelope's diagnostics rather than through a result.
    /// So an empty list here means "this kernel raised none" and not "nothing checked this call" -
    /// and this kernel raises none of its own, which is why the dispatcher hands it `Vec::new()`.
    #[must_use]
    pub fn of(outcome: &kernel::HeatExchangerOutcome, warnings: Vec<Warning>) -> Self {
        Self {
            hot_out_n: outcome.hot_out.n,
            hot_out_z: outcome.hot_out.z.clone(),
            hot_out_p: outcome.hot_out.p,
            hot_out_t: outcome.hot_out.t,
            hot_out_h: joules_per_mole(outcome.hot_out.h.value),
            cold_out_n: outcome.cold_out.n,
            cold_out_z: outcome.cold_out.z.clone(),
            cold_out_p: outcome.cold_out.p,
            cold_out_t: outcome.cold_out.t,
            cold_out_h: joules_per_mole(outcome.cold_out.h.value),
            duty: outcome.duty,
            ntu: outcome.rating.as_ref().map(|rating| rating.ntu),
            effectiveness: outcome.rating.as_ref().map(|rating| rating.effectiveness),
            c_min: outcome.rating.as_ref().map(|rating| rating.c_min),
            c_max: outcome.rating.as_ref().map(|rating| rating.c_max),
            capacity_ratio: outcome.rating.as_ref().map(|rating| rating.capacity_ratio),
            warnings,
        }
    }
}
